ds13/code/modules/mob/living/psychosis_new.dm

```diff
@@ -0,0 +1,106 @@
+#define PSYCHOSIS_EFFECT_LAUGHTER 1
+#define PSYCHOSIS_EFFECT_HITSELF 2
+#define PSYCHOSIS_EFFECT_TRIPPY 3
+#define PSYCHOSIS_EFFECT_WEIRDSCREEN 4
+
+/datum/extension/psychosis
+	name = "Psychotic trauma"
+	expected_type = /mob/living/carbon/human
+	flags = EXTENSION_FLAG_IMMEDIATE
+	var/mob/living/carbon/human/user
+	var/duration = 10 SECONDS
+	var/intensity = 1 //Field to represent the maximum level of trauma that a mob can receive.
+	var/max_intensity = PSYCHOSIS_EFFECT_WEIRDSCREEN //Update this as you code more psychosis effects. Used by the random number generator.
+	var/countdown
+	var/psychosis_timer
+
+/datum/extension/psychosis/New(var/datum/holder, var/_duration, var/_intensity)
+	. = ..()
+	duration = _duration SECONDS
+	intensity = (_intensity <= max_intensity) ? _intensity : max_intensity //Strip off over-intensities that may break things.
+	user = holder
+	start_psychosis()
+
+/**
+
+Method to begin psychosis effects, and start the countdown timer
+
+*/
+
+/datum/extension/psychosis/proc/start_psychosis()
+	if(countdown)
+		deltimer(countdown)
+	to_chat(user, SPAN_WARNING("You start to feel a little bit unhinged..."))
+	countdown = addtimer(CALLBACK(src, /datum/extension/psychosis/proc/finish), duration, TIMER_STOPPABLE) //Countdown to release them from their psychosis.
+	apply_psychosis_effects()
+
+/**
+
+Method to cripple the victim with insanity effects. Calls recursively via timers until finishing.
+
+*/
+
+/datum/extension/psychosis/proc/apply_psychosis_effects()
+	var/effect = rand(PSYCHOSIS_EFFECT_LAUGHTER, intensity) //Update this as required. Only coded in a few as of now
+	switch(effect)
+		if(PSYCHOSIS_EFFECT_LAUGHTER)
+			var/obj/item/organ/external/head/head = user.get_organ("head")
+			if(!istype(head) || !head)
+				return
+			switch(intensity)
+				if(1)
+					user.custom_pain("You suddenly feel an urge to laugh.",0, affecting = LOWER_TORSO)
+				if(2)
+					user.custom_pain("You burst out laughing!",1, affecting = LOWER_TORSO)
+				if(3 to max_intensity)
+					user.custom_pain("You burst out into an uncontrollable fit of laughter!",1, affecting = LOWER_TORSO)
+					user.emote("laugh")
+		if(PSYCHOSIS_EFFECT_HITSELF)
```

Comment:
This could be done much better. Switch the user's intent to harm and make them call attack_hand on themselves